File: services/api-gateway/app/routers/audit.py
"""Audit log read endpoints (tenant-scoped)."""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.auth import require_role
from app.database import get_db
from app.models import AuditEvent, User
from app.schemas import AuditEventResponse

router = APIRouter(prefix="/tenants/{tenant_id}/audit", tags=["audit"])


@router.get("", response_model=list[AuditEventResponse])
def list_audit_events(
    tenant_id: uuid.UUID,
    limit: int = Query(default=50, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_role("PLATFORM_ADMIN", "TENANT_ADMIN")),
):
    """List audit events for a tenant. Platform admins see any tenant; tenant admins see their own."""
    if current_user.role != "PLATFORM_ADMIN" and current_user.tenant_id != tenant_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Access denied")

    return (
        db.query(AuditEvent)
        .filter(AuditEvent.tenant_id == tenant_id)
        .order_by(AuditEvent.created_at.desc())
        .offset(offset)
        .limit(limit)
        .all()
    )

File: services/web-ui/src/api/audit.ts
/**
 * Audit endpoints: /api/tenants/:tenantId/audit
 */

import { apiGet, qs } from "./client";
import type { AuditEventResponse } from "../types";

export interface ListAuditEventsParams {
  limit?: number;
  offset?: number;
}

export function getAuditEvents(
  tenantId: string,
  params: ListAuditEventsParams = {},
): Promise<AuditEventResponse[]> {
  return apiGet<AuditEventResponse[]>(
    `/tenants/${tenantId}/audit${qs({ limit: params.limit, offset: params.offset })}`,
  );
}
